File: apps/backend/app/controllers/internal/expire_controller.rb
module Internal
  class ExpireController < ApplicationController
    def create
      unless authorized?
        return render json: { error: "unauthorized" }, status: :unauthorized
      end

      Transfer.expire_stale!
      ShortLink.expire_stale!
      render json: { ok: true }
    end

    private

    def authorized?
      secret = AnyShare.config.expire_secret
      return true if secret.blank? && !Rails.env.production?

      ActiveSupport::SecurityUtils.secure_compare(bearer_token.to_s, secret.to_s)
    end
  end
end

File: apps/backend/lib/anyshare_web/controllers/internal/expire_controller.ex
defmodule AnyshareWeb.Internal.ExpireController do
  use AnyshareWeb, :controller

  alias Anyshare.Sharing
  alias AnyshareWeb.Plugs.AuthenticateDevice

  def create(conn, _params) do
    if authorized?(conn) do
      :ok = Sharing.expire_stale()
      json(conn, %{ok: true})
    else
      ControllerHelpers.error(conn, :unauthorized, "unauthorized")
    end
  end

  defp authorized?(conn) do
    secret = Application.get_env(:anyshare, :expire_secret)
    environment = Application.get_env(:anyshare, :environment)

    cond do
      not present?(secret) and environment != :prod -> true
      not present?(secret) -> false
      true -> secure_compare(AuthenticateDevice.bearer_token(conn) || "", secret)
    end
  end

  defp secure_compare(left, right) when byte_size(left) == byte_size(right),
    do: :crypto.hash_equals(left, right)

  defp secure_compare(_left, _right), do: false
  defp present?(value), do: is_binary(value) and value != ""
end
